Use ServiceWorker's settings object instead of request's.

yoshisatoyanagisawa opened the pull request:
When we designed the ServiceWorker static routing API, it was designed to be offloading simple things ServiceWorkers do.  With that concept, if we evaluate the cache COEP, it should not be a request's COEP, but to be a ServiceWorker's COEP to make it behave as an offload.

To make it happen, we partially set up the ServiceWorker's settingsObject to be used for the COEP check.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/pull/11.html" title="Last updated on Feb 9, 2024, 7:44 AM UTC (6869fee)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/11/55f9360...6869fee.html" title="Last updated on Feb 9, 2024, 7:44 AM UTC (6869fee)">Diff</a>

yoshisatoyanagisawa committed:
Use ServiceWorker's settings object instead of request's.

sisidovski commented on docs/index.bs:
```diff
@@ -2973,23 +2975,21 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
 
   <section algorithm>
-    <h3 id="run-service-worker-algorithm"><dfn>Run Service Worker</dfn></h3>
+    <h3 id="partial-setup-algorithm"><dfn>Partial Setup</dfn></h3>
```
I feel `Partial Setup` is so not intuitive, how about `Setup Service Worker Global Scope`?

domenic replied:
Naming this operation is tricky.

If I understand correctly, this operation does not represent how things are actually implemented. In the implementation, you will probably not spin up a whole new process (agent), create a ServiceWorkerGlobalScope JavaScript/V8 object, etc. You will just do the minimum amount of work necessary to ensure that the relevant implementation objects for doing a COEP check are ready.

Whereas in the spec, this operation does "most of" the work of starting up the service worker: basically, everything except running the script.

However, this mismatch is OK, because the observable consequences are the same. Nobody could write a test that distinguishes the spec you have written, from some other version of the spec which only does minimal work.

From this perspective, I think @sisidovski's naming suggestion is more accurate on the spec level. But it is less accurate in terms of the implementation level. I think we should use the spec-accurate name, but add a note explaining the situation. Here is a first draft of such a note; you can check if it is accurate and tweak as desired.

> This algorithm does the minimal setup for the service worker that is necessary to create something usable for security checks like CSP and COEP. This specification ensures that this algorithm is called before any such checks are performed.
>
> In specifications, such security checks require creating a `ServiceWorkerGlobalScope`, a relevant settings object, a realm, and an agent. In implementations, the amount of work required might be much less. So, implementations could do less work in their equivalent of this algorithm, and more work in Run Service Worker, as long as the results are observably equivalent. (And in particular, as long as all security checks have the same result.)

yoshisatoyanagisawa replied:
> I feel Partial Setup is so not intuitive, how about Setup Service Worker Global Scope?

Sure.  I renamed this.

> If I understand correctly, this operation does not represent how things are actually implemented. In the implementation, you will probably not spin up a whole new process (agent), create a ServiceWorkerGlobalScope JavaScript/V8 object, etc. You will just do the minimum amount of work necessary to ensure that the relevant implementation objects for doing a COEP check are ready.

Yes.  My plan is executing a part of [EmbeddedWorkerInstance::Start](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/service_worker/embedded_worker_instance.cc;l=230;drc=c0f955cad0bf2a21ec83e95e369eb050c030013a), but not run another process or not fully setup ServiceWorkerGlobalScope.

I also add notes to tell that.  I am not sure how to make them pretty printed, though.

sisidovski commented on docs/index.bs:
```diff
@@ -3013,9 +3013,36 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
-          1. If the [=run CSP initialization for a global object=] algorithm returns "<code>Blocked</code>" when executed upon |workerGlobalScope|, set |startFailed| to true and abort these steps.
+          1. If the [=run CSP initialization for a global object=] algorithm returns "<code>Blocked</code>" when executed upon |workerGlobalScope|, set |setupFailed| to true and abort these steps.
+          1. Set |serviceWorker|'s [=partial start flag=].
+      1. Wait for |serviceWorker|'s [=partial start flag=] is set, or for |setupFailed| to be true.
```
Is [=partial start flag=] or |setupFailed| set asynchronously? Wondering if we need "Wait for~" here.

domenic replied:
Yes, it is set inside the service worker agent, which is in theory in a separate process:

> Let agent be the result of [obtaining a service worker agent](https://html.spec.whatwg.org/multipage/webappapis.html#obtain-a-service-worker-agent), and run the following steps in that context:

(See my other comment about "in theory" vs. in implementation.)

yoshisatoyanagisawa replied:
As @domenic mentioned, since it is different agent, I assume we need to wait for its status change.

domenic commented on docs/index.bs:
```diff
@@ -3151,7 +3178,13 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                      1. If |activeWorker| is not [=running=] or |activeWorker|'s [=partial start flag=] is not set:
+                          1. If the result of running the [=Partial Setup=] algorithm with |activeWorker| is false, then return null.
+
+                          Note: settingsObject for |activeWorker| is ready to use here.
```
Slight rephrase, maybe "If this step succeeds, then |activeWorker|'s [=relevant settings object=] is now ready to use."

yoshisatoyanagisawa replied:
done.

domenic reviewed (commented):
This strategy looks good, but I have some suggestions to increase clarity.

yoshisatoyanagisawa committed:
Reflected comments.

Renamed the algorithm name and the flag name to represents what they do
in the specification.

yoshisatoyanagisawa committed:
Checking how this is rendered.

yoshisatoyanagisawa committed:
cosmetic change

domenic commented on docs/index.bs:
```diff
@@ -2973,23 +2975,25 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+
+      Note: This algorithm returns true if |serviceWorker| has a {{ServiceWorkerGlobalScope}} usable for a CSP check.
+
+      Note: This algorithm does the minimal setup for the service worker that is necessary to create something usable for security checks like CSP and COEP. This specification ensures that this algorithm is called before any such checks are performed.
+      In specifications, such security checks require creating a ServiceWorkerGlobalScope, a relevant settings object, a realm, and an agent. In implementations, the amount of work required might be much less. Therefore, implementations could do less work in their equivalent of this algorithm, and more work in Run Service Worker, as long as the results are observably equivalent. (And in particula ...
```
```suggestion
      <div class="note">
        This algorithm does the minimal setup for the service worker that is necessary to create something usable for security checks like CSP and COEP. This specification ensures that this algorithm is called before any such checks are performed.

        In specifications, such security checks require creating a {{ServiceWorkerGlobalScope}}, a [=relevant settings object=], a [=realm=], and an [=agent=]. In implementations, the amount of work required might be much less. Therefore, implementations could do less work in their equivalent of this algorithm, and more work in [=Run Service Worker=], as long as the results are observably equivalent. (And in particular, as long as all security checks have the same result.)
     </div>
```

I think something like this will make it slightly prettier.

yoshisatoyanagisawa replied:
Thanks.  I have applied this.

yoshisatoyanagisawa committed:
Apply the suggestion.

yoshisatoyanagisawa commented:
@sisidovski Please let me know how it is now.

sisidovski reviewed (approved):
LGTM
